### PR TITLE
8.7.4: Add support for Poetry 1.2.0+'s dependency groups

### PR DESCRIPTION
**Pull request by tiegz**
Poetry 1.2.0 added support for dependency groups in https://github.com/python-poetry/poetry/pull/4260 , e.g.:

```
[tool.poetry.group.test.dependencies]
pytest = "^6.0.0"
pytest-mock = "*"
```

docs: https://github.com/python-poetry/poetry/blob/1.7.1/docs/managing-dependencies.md#dependency-groups

### CODE REVIEW

File: lib/bibliothecary/parsers/pypi.rb [hunk 1]
```diff
@@ -179,10 +187,10 @@ def self.parse_poetry_lock(file_contents, options: {})
         manifest["package"].each do |package|
           # next if group == "_meta"
           group = case package['category']
-                  when 'main'
-                    'runtime'
                   when 'dev'
                     'develop'
+                  else
+                    'runtime'
```
Comment (tiegz) [inline lib/bibliothecary/parsers/pypi.rb:193]: this ensures that `pypoetry.lock` should not get a `type: nil` back

File: spec/fixtures/pyproject.toml [hunk 2]
```diff
@@ -8,9 +8,14 @@ authors = ["Tyrel Souza <tyrel@tidelift.com>"]
 python = "^3.7"
 django = "^3.0.7"
 
+# Old way to define dev deps (<1.2.0)
 [tool.poetry.dev-dependencies]
 pytest = "^5.2"
 
+# New way to define dev deps (>=1.2.0)
+[tool.poetry.group.dev.dependencies]
```
Comment (mikeyoung85) [inline spec/fixtures/pyproject.toml:16]: We should add another group here and test that the type works correctly. Something like `[tool.poetry.group.test.dependencies]`.

File: lib/bibliothecary/parsers/pypi.rb [hunk 3]
```diff
@@ -102,7 +102,15 @@ def self.parse_pyproject(file_contents, options: {})
         # Parse poetry [tool.poetry] deps
         poetry_manifest = file_contents.fetch('tool', {}).fetch('poetry', {})
         deps += map_dependencies(poetry_manifest['dependencies'], 'runtime')
+        # Poetry 1.0.0-1.2.0 way of defining dev deps
         deps += map_dependencies(poetry_manifest['dev-dependencies'], 'develop')
+        # Poetry's 1.2.0+ of defining dev deps
+        poetry_manifest
+          .fetch("group", {})
+          .each_pair do |group_name, obj|
+            group_name = "develop" if group_name == "dev"
```
Comment (mikeyoung85) [inline lib/bibliothecary/parsers/pypi.rb:111]: I'm familiar with toml and how these are getting parsed. What would the expected group names look like. From the Poetry example what would the `group_name` for `[tool.poetry.group.test.dependencies]` resolve to here?

Comment (tiegz) [reply lib/bibliothecary/parsers/pypi.rb:111]: AFAICT it'll use the name that was parsed, so `test` would be the group there. The docs say:

> use a tool.poetry.group.<group> section where <group> is the name of your dependency group (for instance, test):

There's an implicit `main` group that is the default, but we've apparently mapped that to `runtime` until now so I'm setting that as the default in the change below.

Also noteworthy is that anything that's not `main` is considered development semantically, so there's no real distinction between `dev` and `test` to Poetry other than the names:

>  Dependency groups, other than the implicit main group, must only contain dependencies you need in your development process. Installing them is only possible by using Poetry.

And FWIW, I'd prefer to use "development" here to be consistent with the other parsers, but there are 3 uses of "develop" in this file so I'm hesitant to break anything for consumers of the gem : \

**Review by mikeyoung85 — APPROVED** [review]

**Pull request merged**